mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,311 @@ def load_model(model_uri):
+        we can query the eval dataset name and the model id via the `y_pred` argument instance.
+        """
+        value = (eval_dataset_name, self.get_model_id(model))
+        prediction_result_id = id(predict_result)
+        self._pred_result_id_to_dataset_name_and_model_id[prediction_result_id] = value
+
+        def clean_id(id_):
+            _get_autolog_training_status()._pred_result_id_to_dataset_name_and_model_id \
+                .pop(id_, None)
+
+        # When the `predict_result` object being GCed, its ID may be reused, so register a finalizer
+        # to clear the ID from the dict for preventing wrong ID mapping.
+        weakref.finalize(predict_result, clean_id, prediction_result_id)
+
+    def register_metric_api_call(self, metric_fn, call_pos_args, call_kwargs):
+        """
+        Given a metric api call (include the called metric function, and call arguments)
+        Register the call information (arguments dict) into the `metric_api_call_arg_dict_list_map`
+        and return a tuple of (is_ok, model_id, metric_key)
+        - metric key format is: {metric_name}_on_{eval_dataset_name}
+          metric_name is generated by metric function name, if there're multiple calls with
+          different call arguments, then appending index to the name.
+        """
+        call_arg_list = list(call_pos_args) + list(call_kwargs.values())
+
+        dataset_id_list = self._pred_result_id_to_dataset_name_and_model_id.keys()
+        print(f'DGB: get_eval_dataset_name: # 1')
+
+        dataset_name = None
+        model_id = None
+
+        # Note: some metric API the arguments is not like `y_true`, `y_pred`
+        #  e.g.
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html#sklearn.metrics.silhouette_score
+        for arg in call_arg_list:
+            if arg is not None and not np.isscalar(arg) and id(arg) in dataset_id_list:
+                print(f'DGB: get_eval_dataset_name: # 3')
+                dataset_name, model_id = self._pred_result_id_to_dataset_name_and_model_id[id(arg)]
+                break
+
+        if dataset_name is None or model_id is None:
+            return False, None, None
+
+        run_id = self._model_id_to_run_id_map.get(model_id, None)
+        if run_id is None:
+            return False, None, None
+
+        call_arg_dict = _extract_metric_api_call_arg_dict(metric_fn, call_pos_args, call_kwargs)
+        metric_name = metric_fn.__name__
+
+        registered_call_arg_dict_list = \
+            self._metric_api_call_arg_dict_list_map[model_id][metric_name]
+
+        index = len(registered_call_arg_dict_list)
+        for i, call_arg_dict_i in enumerate(registered_call_arg_dict_list):
+            if call_arg_dict == call_arg_dict_i:
+                index = i
+                break
```

Comment:
I think we should just write a new MLflow metric each time a sklearn metric function is called.

i.e:

```
accuracy_score(test_X, test_y) # should produce `accuracy_score_on_test_X`
accuracy_score(test_X, test_y) # should produce `accuracy_score_on_test_X_2`.
```

This is simpler than trying to determine whether or not a metric API was called twice. It's also easier for users to reason about because they can just look at how many metric API calls were included in the source code, rather than filtering out duplicates in order to understand the indexing.

If we make this change, we can get rid of a lot of complex logic in `_exactract_metric_api_call_arg_dict`. We'd just need to maintain a mapping from `metric_key` to the count of how many times it's been used.

Reply:
Sounds good.
So for each run_id, create a global metric_call_index starting from 1, when detect one metric API called, the logged metric key will be appended with the index, right ?

then shall we still log the additional metric_info artifact ? (log the metric arguments value for each metric call). I suggest to still log it.